wip: Rework dependencies, extend expression AST

afiore opened the pull request.

Commit:
rework build dependencies

also re-implement encoders using macros only

Commit:
Extend expression with Property and PropertyF

Commit:
rework literal implementation

also start implementing Template and Parameter

Commit:
wip

Commit:
Implement parameters

Commit:
remove type parameter

Commit:
wip

afiore commented on cf/src/main/scala/com/timeout/scalacloudformation/Encoding.scala:
```diff
@@ -43,30 +50,60 @@ object Encoding {
+    val awsEnum = implicitly[Enum[Parameter.AwsType]]
+
+    Encoder.instance[Parameter.DataType] { dt =>
+      val s = dt match {
```
I tried using `scala.reflect`'s `WeakTypedTag#tpe.typeSymbol.name`, but unfortunately it does return fairly weird string values such us "ListLtNumberGt". So this ends up being the most pragmatic way in this case

Commit:
add common properties

Commit:
Complete implementation of template

afiore commented on cf/src/main/scala/com/timeout/scalacloudformation/package.scala:
```diff
@@ -0,0 +1,54 @@
+    def jsonEncode: Json
+    def DependsOn: Option[CfExp[String]]
+    def DeletionPolicy: Option[CfExp[DeletionPolicy]]
+    def CreationPolicy: Option[CfExp[CreationPolicy]]
```
Looks like `CreationPolicy` is supported only by [three resource|http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-creationpolicy.html]: 
- AWS::AutoScaling::AutoScalingGroup
- AWS::EC2::Instance
- AWS::CloudFormation::WaitCondition.

Perhaps we should remove it from this trait and add it in CodeGen, as I have done for `UpdatePolicy`

afiore commented on cf/src/main/scala/com/timeout/scalacloudformation/package.scala:
```diff
@@ -0,0 +1,54 @@
+    def fqn: String
+    def logicalId: String
+    def jsonEncode: Json
+    def DependsOn: Option[CfExp[String]]
```
[DependsOn](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-dependson.html) would probably be best typed as `CfExp[ResourceRef]`, but it is serialised as a plain string

ostapneko replied:
Not sure `DependsOn` can take a `CfExp`. Instead, we could have `def DependsOn: Option[Resource]`

afiore commented on cf/src/test/scala/com/timeout/scalacloudformation/ParameterTest.scala:
```diff
@@ -0,0 +1,124 @@
+  def json(s: String): Json =
+    parser.parse(s).right.get
+
+  def dropNullKeys(j: Json): Json ={
```
This is admittedly gross, but it's test code :)

afiore commented on macros/src/main/scala/com/timeout/CodeGen.scala:
```diff
@@ -64,14 +75,39 @@ class CodeGen(conf: CodeGen.Config) {
+               logicalId -> Json.obj(
+                 "Type" -> Json.fromString(fqn),
+                 "Properties" -> Json.obj(..$jsonFields)
+               )
```
Actually, `DependsOn`, `DeletionPolicy`, `CreationPolicy`, and `UpdatePolicy`, go in here!

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/CfExp.scala:
```diff
@@ -1,22 +1,30 @@
+    implicit val jsonLit: IsLit[Json] = new IsLit[Json] {}
+    implicit val durationLit: IsLit[Duration] = new IsLit[Duration] {}
+    implicit def propertyLit[T <: ResourceProperty]: IsLit[T] = new IsLit[T] {}
+    implicit def listLit[A: IsLit]: IsLit[List[A]] = new IsLit[List[A]]{}
```
Only lists of strings are permitted according to their spec
http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resources-section-structure.html

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/CfExp.scala:
```diff
@@ -1,22 +1,30 @@
+
+  case class Lit[T: IsLit](value: T) extends E[T]
 
   /** The return value for ref depends on the resource
```
We are missing parameter refs

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/Parameter.scala:
```diff
@@ -0,0 +1,82 @@
+object Parameter {
+  sealed trait AwsType
+  object AwsType {
+    case object `AWS::EC2::Instance` extends AwsType
```
Are you sure about this type? The spec only talks about `AWS::EC2::Instance::Id`, which you added below

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/Parameter.scala:
```diff
@@ -0,0 +1,82 @@
+    case object `List<Number>` extends DataType
+    case object CommaDelimitedList extends DataType
+    case class AwsType(tpe: Parameter.AwsType) extends DataType
+    case class AwsListType(tpe: Parameter.AwsType) extends DataType
```
I think this should be called `AwsTypeList` instead. Also, I don't know if it's accurate, as only a subset of the AWS types are available as list items

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/Parameter.scala:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  case class Str(logicalId: String,
+                 MaxLength: Option[Int] = None,
```
This should output strings

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/Parameter.scala:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  case class Number(logicalId: String,
+                    MaxValue: Option[Double] = None,
```
ditto

ostapneko replied:
Don't we have `AllowedValues` for numbers?

afiore replied:
Looks like we don't

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/Parameter.scala:
```diff
@@ -0,0 +1,82 @@
+                 Description: Option[String] = None,
+                 NoEcho: Option[Boolean] = None,
+                 AllowedValues: Option[Set[String]] = None,
+                 AllowedPattern: Option[String] = None,
```
I think we are missing `ConstraintDescription`

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/package.scala:
```diff
@@ -0,0 +1,54 @@
+import io.circe.Json
+import enum.Enum
+
+package object scalacloudformation {
```
We could put all that in a `ResourceAttributes` object, to map with the doc's terminology

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/package.scala:
```diff
@@ -0,0 +1,54 @@
+                          AutoScalingRollingUpdate: Option[AutoScalingRollingUpdate] = None,
+                          AutoScalingScheduledAction: Option[AutoScalingScheduledAction] = None) extends ResourceProperty
+
+  trait Resource extends HasLogicalId {
```
I don't think that Resource attributes can take `CfExp`:

> You can use intrinsic functions only in specific parts of a template. Currently, you can use intrinsic functions in resource properties, outputs, metadata attributes, and update policy attributes.

ostapneko replied:
We can add `def Metadata: Option[Json]`

ostapneko replied:
Also, it looks like `def UpdatePolicy: Option[UpdatePolicy]` is missing

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/Template.scala:
```diff
@@ -0,0 +1,12 @@
+
+case class Template(Description: Option[String] = None,
+                    Metadata: Option[Json] = None,
+                    Mappings: Option[Json] = None,
```
I implemented this as a simple `Mappings: Map[String, Map[String, String]]` since it's always a two-levels map of strings, without expression...
http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/mappings-section-structure.html

afiore replied:
are you sure is always two levels of depth?

ostapneko replied:
Yes, and it corresponds to the arity of FindInMap

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-findinmap.html

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/Encoding.scala:
```diff
@@ -43,30 +55,110 @@ object Encoding {
+
+  implicit val template: Encoder[Template] = Encoder.instance { t =>
+    Json.obj(
+      "AWSTemplateFormatVersion" -> Json.fromString("2010-09-09"),
```
Why do we fix it?

afiore replied:
Because is the only version defined so far by the specification

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/Encoding.scala:
```diff
@@ -43,30 +55,110 @@ object Encoding {
+      "Description" -> o.Description.asJson,
+      "Value" -> o.Value.asJson,
+      "Condition" -> o.Condition.map(_.logicalId).asJson,
+      "Export" -> o.Export.asJson //TODO: support cross-stack output
```
It looks like the name is always wrapped into a `Name` key:

```javascript
"Outputs" : {
  "Logical ID" : {
    "Description" : "Information about the value",
    "Value" : "Value to return",
    "Export" : {
      "Name" : "Value to export"
    }
  }
}
```

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html

ostapneko commented on cf/src/main/scala/com/timeout/scalacloudformation/Output.scala:
```diff
@@ -0,0 +1,7 @@
+package com.timeout.scalacloudformation
+
+case class Output(logicalId: String,
+                  Value: CfExp[String],
```
Apparently you cannot have expressions for `Value`

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html

Commit:
Implement feedback from PR

- Implement ParamterRef
- Correct encoding for Resource and Output
- Distinguish between Parameter.Integer and Parameter.Double
- Redefine mappings as a Map[String, Map[String, String]]

ostapneko commented:
👍 after rebase
